modules/amm-executor/src/main/scala/org/ergoplatform/dex/executor/amm/processes/Executor.scala

```diff
@@ -48,34 +51,50 @@ object Executor {
     }
 
   final private class Live[
-    F[_]: Monad: Evals[*[_], G],
-    G[_]: Monad: Logging: TraceId.Local: Clock: Catches
+    F[_]: Monad: Evals[*[_], G]: Defer: SemigroupK: ParFlatten,
+    G[_]: Monad: Logging: Catches: Timer
   ](conf: ExecutionConfig)(implicit
     orders: CFMMCircuit[F, G],
+    executedOrders: CFMMHistConsumer[F, G],
+    backlog: CFMMBacklog[G],
     service: Execution[G],
     errParser: TxSubmissionErrorParser
   ) extends Executor[F] {
 
     def run: F[Unit] =
+      emits(
+        List(
+          addToBacklog,
+          executeOrders,
+          dropExecuted
+        )
+      ).parFlattenUnbounded
+
+    def addToBacklog: F[Unit] =
       orders.stream
-        .evalMap { rec =>
-          service
-            .executeAttempt(rec.message)
-            .handleWith[Throwable](e => warnCause"Order execution failed fatally" (e) as none[CFMMOrder])
-            .local(_ => TraceId.fromString(rec.message.id.value))
-            .tupleLeft(rec)
+        .evalTap(orderRec => backlog.put(orderRec.message))
+        .evalMap(_.commit)
+
+    def dropExecuted: F[Unit] =
+      executedOrders.stream
+        .evalTap { rec =>
+          rec.message.traverse(order => backlog.drop(order.order.id))
         }
-        .flatTap {
-          case (_, None) => unit[F]
-          case (_, Some(order)) =>
-            eval(now.millis) >>= {
-              case ts if ts - order.timestamp < conf.orderLifetime.toMillis =>
-                eval(warn"Failed to execute $order. Going to retry.") >>
-                  orders.retry((order.id -> order).pure[F])
-              case _ =>
-                eval(warn"Failed to execute $order. Order expired.")
-            }
+        .evalMap(_.commit)
```

Comment:
For Kafka it's inefficient to commit messages by one

Reply:
We can't batch orders, because of it will slow down execution

Reply:
How does it affect?

Reply:
We can execute asap but make commit later

Reply:
"We can execute asap but make commit later" - how?

Reply:
The simplest way is to have a count value of processed orders and when the count is greater than *value from config*, make a commit

Reply:
Remember that we have commit only the latest one message not everyone so we can simply skip all prev `Committable`

Reply:
in such case we will retry orders execution, if some err occurs. Do you think it will be correct?

Reply:
I'm not sure about it. The same problem exists now: if the error occurs in timeline between `service.executeAttempt(order)` and (`backlog.suspend(order)` or `backlog.checkLater(order)`) we will lost order at all. So, they both are good questions, need @oskin1 here

Reply:
Backlog should be designed in a way that preserves at-least-once processing guarantee.

Reply:
I think we solved this in Haskell implementation.

Reply:
So, with that information it's okay to process orders the way I suggested